### ISSUE TITLE
Bug: RuleTester doesn't fix properly when there are multiple fixes on the same line

### ISSUE THREAD

**Issue by StyleShit** [typescript-eslint/typescript-eslint#8251]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

rule-tester

### Repro Code
See: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/space-infix-ops.test.ts#L1021-L1049
```TypeScript
{
  invalid: [
    {
      code: `
        type Test=|string|(((() => void)))|string;
      `,
      output: `
        type Test = |string | (((() => void))) | string;
      `,
      errors: [
        {
          messageId: 'missingSpace',
          column: 18,
          line: 2,
        },
        {
          messageId: 'missingSpace',
          column: 19,
          line: 2,
        },
        {
          messageId: 'missingSpace',
          column: 26,
          line: 2,
        },
        {
          messageId: 'missingSpace',
          column: 43,
          line: 2,
        },
      ],
    },
  ],
}
```

### Description

Splits from this thread:
https://github.com/typescript-eslint/typescript-eslint/pull/8065#discussion_r1429216128

The code should be fixed properly when there are multiple fixes on the same line.
In the repro code, the fix should actually be
```TS
type Test = | string | (((() => void))) | string;
```

And in the thread, it should've been fixed regardless of me adding space as a workaround
(you can see that I've changed it from `${'a'}${'b'}` to `${'a'} ${'b'}`)

Opened a [draft PR](https://github.com/typescript-eslint/typescript-eslint/pull/8252) to start an investigation.
Seems like there are some actual wrong fixes in the tests.

**Comment by StyleShit** [comment]
Based on this [random test](https://github.com/eslint/eslint/blob/c893bc0bdf1bca256fbab6190358e5f922683249/tests/lib/rules/space-in-parens.js#L317-L327), and after playing a little bit with the original `source-code-fixer`, it seems to work fine.

I'm not sure why it doesn't work here as I expect it to, while it _does_ work as expected in eslint.
I feel like I'm missing something here...

**Comment by bradzacher** [comment]
@StyleShit 
You're likely expecting that the rule tester "exhaustively runs fixers" for each test case.
However it does not - the rule tester does exactly one fix pass:

https://github.com/eslint/eslint/blob/1f3744278433006042b8d5f4e9e1e488b2bbb011/lib/rule-tester/rule-tester.js#L760-L770

OTOH with ESLint itself when you run with `--fix` it'll lint the file and apply fixes up to 10 times in a row (or until the file is no longer fixed, whichever is first).
This is how it handles cases of overlapping fixes.
https://github.com/eslint/eslint/blob/1f3744278433006042b8d5f4e9e1e488b2bbb011/lib/linter/linter.js#L2160-L2186

**Comment by StyleShit** [comment]
@bradzacher 
Well, not exactly
The test fixer takes all of the fixes from left to right and fixes them in a single pass:
https://github.com/eslint/eslint/blob/1f3744278433006042b8d5f4e9e1e488b2bbb011/lib/linter/source-code-fixer.js#L121-L133

So I expect it to do the same here, but for some reason it doesn't and I have no idea why (because as I mentioned above, it _does_ work in eslint 😓)

**Comment by bradzacher** [comment]
Sorry by a pass I mean "calculate the lint messages for the entire file then apply the fixers from left to right, dropping the ones that conflict with already applied fixes".

The rule tester does one "pass" and the CLI does up to 10 "passes". 

If you turn on debug logging in the ESLint CLI (`DEBUG=*`) I would assume that you'll see eslint logs out that it does two passes on the code. Which would be why it "works" in the CLI.

**Comment by StyleShit** [comment]
Yeah, I understood what you meant by a "pass", you (and the code) were pretty clear before :) 

What I'm saying is that the fixer works in eslint's tests (not talking about the CLI at all), when there are multiple fixes in the same line, even when they are close to each other (as long they're not conflicting of course), but for some reason it doesn't work here (which is why I had to add a space between 2 lint messages/fixes in the test mentioned above)

I'm hoping that this time I'm clearer

**Comment by bradzacher** [comment]
If you run the debugger on a lint run you can see exactly what ESLint does from the CLI.

1) input code `type Test=|string|(((() => void)))|string;`
2) run rule on code and get 4 messages back
3) apply fixers from messages on code and get `type Test = |string | (((() => void))) | string;`
4) as fixers were applied - do another pass
5) input code `type Test = |string | (((() => void))) | string;`
6) run rule on code and get 1 message back
7) apply fixers from messages on code and get `type Test = | string | (((() => void))) | string;`
8) as fixers were applied - do another pass
9) input code `type Test = | string | (((() => void))) | string;`
10) run rule on code and get 0 messages back
11) apply fixers from messages on code (no changes as no messages)
11) no fixers were applied - exit

So there are 3 total passes that are done here from the CLI - in particular you can see that the first pass fixes it to the same state you are seeing from the rule tester. This is to be expected because the rule tester does exactly 1 pass.

The reason this happens is that the first two reports overlap - their ranges are 1:10-1:11 and 1:11-1:12. Because the end of one range === the start of the next - the second range is overlapping and thus the fix is dropped.

**Comment by bradzacher** [comment]
Filed https://github.com/eslint/eslint/issues/18007 to try and improve this state - but currently this is "working as intended"
